emma-language/src/main/scala/org/emmalanguage/api/DataBag.scala

```diff
@@ -311,6 +316,34 @@ trait DataBag[A] extends Serializable {
           }
       })._1
   }
+
+  // -----------------------------------------------------
+  // equals, hashCode and toString
+  // -----------------------------------------------------
+
+  override def equals(o: Any) = o match {
+    case that: DataBag[A] => {
+      val thisVals = fetch()
+      val thatVals = that.fetch()
+      (thisVals.size == thatVals.size) &&
```

Comment:
Sorry, I missed the multiplicities. But still, it's possible to implement `minus`, just a little more involved:

First, we convert both bags to the representation `(value, multiplicity)`, e.g. the bag
```
a
a
b
```
becomes
```
(a,2)
(b,1)
```
This is possible with a `groupBy` and then `map size`.
Then we can use an inner join to check that the multiplicities of those elements that appear in both bags match, and then use exists to check for values that appear in only one of them.